File: ObjC/InfiniteViewUseCollectionV/ALInfiniteView/ALInfiniteView.m
//
//  ALInfiniteView.m
//
//
//  Created by April on 2017/7/12.
//  Copyright © 2017年 April. All rights reserved.
//

#import "ALInfiniteView.h"
#import "ALInfiniteViewCell.h"

static NSString *ALInfiniteCellIdentifier = @"ALInfiniteCellIdentifier";

@interface ALInfiniteView () <UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, UIScrollViewDelegate>

@property (weak, nonatomic) IBOutlet UICollectionView *ALInfiniteCollectionView;
@property (weak, nonatomic) IBOutlet UIPageControl *pageControl;
@property (weak, nonatomic) IBOutlet UIView *contentView;

@property (strong, nonatomic) NSArray *originItems;
@property (strong, nonatomic) NSArray *infiniteItems;
@property (strong, nonatomic) NSTimer *autoScrollTimer;

@end

@implementation ALInfiniteView

- (instancetype)initWithFrame:(CGRect)frame owner:(id)owner {
    
    self = [[[NSBundle mainBundle] loadNibNamed:@"ALInfiniteView" owner:owner options:nil] objectAtIndex:0];
    
    if (self) {
        self.frame = frame;
        [self initialView];
    }
    
    return self;
}


/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

#pragma mark - initail

- (void)initialView {
    
    UINib *cellNib = [UINib nibWithNibName:@"ALInfiniteViewCell" bundle:nil];
    [_ALInfiniteCollectionView registerNib:cellNib forCellWithReuseIdentifier:ALInfiniteCellIdentifier];
    _ALInfiniteCollectionView.delegate = self;
    _ALInfiniteCollectionView.dataSource = self;
    
    _ALInfiniteCollectionView.scrollsToTop = false;
}

- (void)createInfiniteItems:(NSArray *)contentItems {
    _originItems = contentItems;
    
    _infiniteItems = [self createInfiniteDataWithOriginialData:_originItems];
    [_ALInfiniteCollectionView reloadData];
    
    _pageControl.numberOfPages = _originItems.count;
}

- (NSArray *)createInfiniteDataWithOriginialData:(NSArray *)orginialData {
    if (orginialData.count <= 1) {
        return orginialData;
    }
    
    id firstItem = [orginialData firstObject];
    id lastItem = [orginialData lastObject];
    
    NSMutableArray *infiniteArray = [orginialData mutableCopy];
    
    [infiniteArray insertObject:lastItem atIndex:0];
    [infiniteArray addObject:firstItem];
    
    return infiniteArray;
}

- (void)scrollToVisibleFirstItem {
    if (_originItems.count <= 1) {
        return;
    }
    
    if (_contentView.isHidden) {
        [_contentView setHidden:NO];
    }
    
    [_ALInfiniteCollectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:1 inSection:0] atScrollPosition:UICollectionViewScrollPositionLeft animated:NO];
    
    [_pageControl setCurrentPage:0];
}

#pragma mark - pageControl

- (IBAction)pageControlValueChanged:(id)sender {
    CGFloat pageWidth = _ALInfiniteCollectionView.bounds.size.width;
    CGPoint scrollTo = CGPointMake(pageWidth * (_pageControl.currentPage + 1), 0);
    
    [_ALInfiniteCollectionView setContentOffset:scrollTo animated:YES];
}

- (void)setPageControlCurrentPage:(UIScrollView *)scrollView {
    CGFloat pageWidth = _ALInfiniteCollectionView.bounds.size.width;
    CGFloat contentOfffsetX = scrollView.contentOffset.x;
    
    // The first page actually is second page in collection view.
    _pageControl.currentPage = ((contentOfffsetX / pageWidth) - 1);
}

#pragma mark - AutoScroll

- (void)startAutoScroll {
    if (_infiniteItems.count <= 1) {
        return;
    }
    
    if (_autoScrollTimer) {
        [self stopAutoScroll];
    }
    
    [self scrollToVisibleFirstItem];
    
    _isAutoScrolling = YES;
    _autoScrollTimer = [NSTimer scheduledTimerWithTimeInterval:5.0 target:self selector:@selector(autoScrollAction) userInfo:nil repeats:YES];
}

- (void)stopAutoScroll {
    if(_autoScrollTimer) {
        _isAutoScrolling = NO;
        [_autoScrollTimer invalidate];
        _autoScrollTimer = nil;
    }
}

- (void)autoScrollAction {
    CGFloat offset = _ALInfiniteCollectionView.contentOffset.x + _ALInfiniteCollectionView.bounds.size.width;
    [_ALInfiniteCollectionView setContentOffset:CGPointMake(offset, 0) animated:YES];
}

#pragma mark - CollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return collectionView.bounds.size;
}

#pragma mark - CollectionView DataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return _infiniteItems.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    NSDictionary *itemInfo = _infiniteItems[indexPath.row];
    
    ALInfiniteViewCell *infiniteCell = (ALInfiniteViewCell *)[collectionView dequeueReusableCellWithReuseIdentifier:ALInfiniteCellIdentifier forIndexPath:indexPath];

    infiniteCell.tag = indexPath.row;
    infiniteCell.NameLabel.text = itemInfo[@"Name"];
    infiniteCell.contentImageView.backgroundColor = itemInfo[@"Color"];
    
    return infiniteCell;
}

#pragma mark -  CollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    if ([_delegate respondsToSelector:@selector(ALInfiniteViewDidSelectedItem:selectedIndex:)]) {
        [_delegate ALInfiniteViewDidSelectedItem:self selectedIndex:indexPath.row];
    }
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath {
    if (_autoScrolled && _autoScrollTimer == nil) {
        [self startAutoScroll];
    } else {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            [self scrollToVisibleFirstItem];
        });
    }
}

#pragma mark - ScrollViewDelegate
//reference for someone, I forget...
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    static CGFloat lastContentOffsetX = FLT_MIN;
    
    //Ignore the first time scroll
    if (FLT_MIN == lastContentOffsetX) {
        lastContentOffsetX = scrollView.contentOffset.x;
    }
    
    CGFloat currentOffsetX = scrollView.contentOffset.x;
    CGFloat currentOffsetY = scrollView.contentOffset.y;
    
    CGFloat pageWidth = scrollView.bounds.size.width;
    CGFloat offset = pageWidth * (_infiniteItems.count - 2);
    
    //the first page(showing the last item) is visible and user is still scrolling to the left
    if (currentOffsetX < pageWidth && lastContentOffsetX > currentOffsetX) {
        lastContentOffsetX = currentOffsetX + offset;
        scrollView.contentOffset = CGPointMake(lastContentOffsetX, currentOffsetY);
        
    } else if (currentOffsetX > offset && lastContentOffsetX < currentOffsetX) {
        // the last page (showing the first item) is visible and user is still scrolling to the right
        lastContentOffsetX = currentOffsetX - offset;
        scrollView.contentOffset = (CGPoint){lastContentOffsetX, currentOffsetY};
    } else {
        lastContentOffsetX = currentOffsetX;
    }
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    [self setPageControlCurrentPage:scrollView];
}

- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView {
    [self setPageControlCurrentPage:scrollView];
}



@end

File: Swift/infiniteViewUseCollectionViewSwift/infiniteViewUseCollectionViewSwift/ALInfiniteView/ALInfiniteView.swift
//
//  ALInfiniteView.swift
//  infiniteViewUseCollectionViewSwift
//
//  Created by April on 2017/7/18.
//  Copyright © 2017年 April. All rights reserved.
//

import UIKit

let ALInfiniteCellIdentifier = "ALInfiniteCellIdentifier"

protocol ALInfiniteViewDelegate {
    func alInfiniteViewDidSelectedItem(infiniteView: ALInfiniteView, selectedIndex: Int)
}

class ALInfiniteView: UIView {
    
    @IBOutlet weak var backgroudImageView: UIImageView!
    @IBOutlet weak var contentView: UIView!
    
    @IBOutlet var alInfiniteView: UIView!
    @IBOutlet weak var ALInfiniteCollectionView: UICollectionView!
    @IBOutlet weak var pageControl: UIPageControl!
    
    var originItems: Array<Any>?
    var infiniteItems = [Any]()
    var autoScrollTimer: Timer?
    
    var delegate: ALInfiniteViewDelegate?
    
    override init(frame: CGRect) { //code
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) { //IB
        super.init(coder: aDecoder)
        setup()
        fatalError("init(coder:) has not been implemented")
    }
    
    
    func setup() {
        Bundle.main.loadNibNamed("ALInfiniteView", owner: self, options: nil)
        addSubview(alInfiniteView)
        alInfiniteView.frame = self.bounds
        alInfiniteView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        
        setupCollectionView()
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */
    
    //MARK: - View Setup
    
    private func setupCollectionView () {
        
        let cellNib = UINib.init(nibName: "ALInfiniteViewCell", bundle: nil)
        ALInfiniteCollectionView.register(cellNib, forCellWithReuseIdentifier: ALInfiniteCellIdentifier)
        
        ALInfiniteCollectionView.delegate = self
        ALInfiniteCollectionView.dataSource = self
        ALInfiniteCollectionView.scrollsToTop = false
    }
    
    func setInfiniteViewItems (orginItems: [[String: Any]]) {
        self.originItems = orginItems
        
        infiniteItems = createInfiniteItems(originItems: orginItems)
        pageControl.numberOfPages = originItems!.count
    }
    
    //MARK: - Infinite Items

    private func createInfiniteItems(originItems: [[String: Any]]) -> [[String: Any]]{
        if originItems.count <= 1 {
            return originItems
        }
        
        let firstItem = originItems.first
        let lastItem = originItems.last
        
        var infiniteItems = originItems
        
        infiniteItems.insert(lastItem!, at: 0)
        infiniteItems.append(firstItem!)
        
        return infiniteItems
    }
    
    //MARK: - Scroll Action
    func scrollToVisibleFirstItem() {
        guard originItems!.count > 1 else {
            return
        }
        
        ALInfiniteCollectionView.scrollToItem(at: IndexPath(row: 1, section: 0) , at: .left, animated: false)
        pageControl.currentPage = 0
    }

    //MARK: - PageControl
    @IBAction func pageControlValueChanged(_ sender: Any) {
        let pageWidth = ALInfiniteCollectionView.bounds.size.width
        let scrollTo = CGPoint(x: (pageWidth * (CGFloat)(pageControl.currentPage + 1)), y: 0)
        
        ALInfiniteCollectionView.setContentOffset(scrollTo, animated: true)
    }
    
    func setPageCotrolTheCurrentPage(scrollView: UIScrollView) {
        let pageWidth = ALInfiniteCollectionView.bounds.size.width
        let contentOffsetX = scrollView.contentOffset.x
        
        pageControl.currentPage = ((Int)(contentOffsetX / pageWidth) - 1)
    }
    
    //MARK: - AutoScroll
    
    func startAutoScroll() {
        guard infiniteItems.count > 1 else {
            return
        }
        
        if autoScrollTimer != nil {
            stopAutoScroll()
        }
        
        scrollToVisibleFirstItem()
        
        autoScrollTimer = Timer(timeInterval: 5.0, target: self, selector: #selector(autoScrollAction), userInfo: nil, repeats: true)
        
    }
    
    func autoScrollAction() {
        let offset = ALInfiniteCollectionView.contentOffset.x + ALInfiniteCollectionView.bounds.size.width
        ALInfiniteCollectionView.setContentOffset(CGPoint(x: offset, y: 0), animated: true)
    }
    
    func stopAutoScroll() {
        if autoScrollTimer != nil {
            autoScrollTimer?.invalidate()
            autoScrollTimer = nil
        }
    }
}

extension ALInfiniteView: UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        delegate?.alInfiniteViewDidSelectedItem(infiniteView: self, selectedIndex: indexPath.item)
    }
}

extension ALInfiniteView: UICollectionViewDataSource {

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return infiniteItems.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let itemInfo: Dictionary = infiniteItems[indexPath.row] as! [String: Any]
        
        let infiniteCell = collectionView.dequeueReusableCell(withReuseIdentifier: ALInfiniteCellIdentifier, for: indexPath) as! ALInfiniteViewCell
        
        infiniteCell.tag = indexPath.item
        infiniteCell.nameLabel.text = itemInfo["Name"] as? String
        infiniteCell.contentImageView.backgroundColor = itemInfo["Color"] as? UIColor
        
        return infiniteCell
    }
}

extension ALInfiniteView: UICollectionViewDelegateFlowLayout {
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return collectionView.bounds.size
    }
}

extension ALInfiniteView: UIScrollViewDelegate {

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
       
        struct lastOffsetX {
            static var lastContentOffsetX = CGFloat.leastNormalMagnitude
        }

        // Ignore the first time scroll
        if lastOffsetX.lastContentOffsetX == CGFloat.leastNormalMagnitude {
            lastOffsetX.lastContentOffsetX = scrollView.contentOffset.x
        }
        
        let currentOffsetX = scrollView.contentOffset.x
        let currentOffsetY = scrollView.contentOffset.y
        
        let pageWidth = scrollView.bounds.size.width
        let offset = pageWidth * CGFloat((infiniteItems.count - 2))
        
        //the first page (showing the last item) is visible and user is still scrolling to the left
        if currentOffsetX < pageWidth && lastOffsetX.lastContentOffsetX > currentOffsetX {
            lastOffsetX.lastContentOffsetX = currentOffsetX + offset
            scrollView.contentOffset = CGPoint(x: lastOffsetX.lastContentOffsetX, y: currentOffsetY)
        } else if currentOffsetX > offset && lastOffsetX.lastContentOffsetX < currentOffsetX {
            //the last page (showing the first item) is visible and use is still scrolling to the right
            lastOffsetX.lastContentOffsetX = currentOffsetX - offset
            scrollView.contentOffset = CGPoint(x: lastOffsetX.lastContentOffsetX, y: currentOffsetY)
        } else {
            lastOffsetX.lastContentOffsetX = currentOffsetX
        }
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        setPageCotrolTheCurrentPage(scrollView: scrollView)
    }
    
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        setPageCotrolTheCurrentPage(scrollView: scrollView)
    }
}
